src/core/QueryManager.ts

```diff
@@ -438,20 +477,38 @@ export class QueryManager<TStore> {
                 returnPartialData: true,
               });
 
-              if (diff.complete && !(isExecutionPatchIncrementalResult(result))) {
-                result = { ...result, data: diff.result };
+              if (diff.complete) {
+                result = { ...result as FetchResult, data: diff.result };
+                if ('incremental' in result) {
+                  delete result.incremental;
+                }
+                if ('hasNext' in result) {
+                  delete result.hasNext;
+                }
               }
             }
 
-            update(cache, result, {
-              context: mutation.context,
-              variables: mutation.variables,
-            });
+            // If we've received the whole response,
+            // either a SingleExecutionResult or the final ExecutionPatchResult,
+            // call the update function.
+            if (isFinalResult) {
+              update(cache, result, {
+                context: mutation.context,
+                variables: mutation.variables,
+              });
+            }
           }
 
           // TODO Do this with cache.evict({ id: 'ROOT_MUTATION' }) but make it
           // shallow to allow rolling back optimistic evictions.
-          if (!skipCache && !mutation.keepRootFields) {
+          if (
+            !skipCache &&
+            !mutation.keepRootFields &&
+            (
+              !isExecutionPatchResult(result) ||
+              (isExecutionPatchIncrementalResult(result) && !result.hasNext)
```

Comment:
```suggestion
              isFinalResult
```

Now that you've extracted the variable, could you use it here?

Reply:
Oops! Updated :)